### ISSUE TITLE
Allow setting application namespace from ApplicationSet template

### ISSUE THREAD

**Issue by waaghals** [argoproj/argo-cd#11104]
# Summary

With the introduction of namespaced applications in #9755 we can have applications in different namespaces.
But the applications generated from an ApplicationSet always use the namespace from the ApplicationSet.

ApplicationSet should allow setting `.spec.template.metadata.namespace` for the generated application.

# Motivation

When deploying the same repository/directory to multiple clusters, one could use ApplicationSets to automatically generate applications for each cluster.
The names of these applications can become rather large as they might include both the application name and cluster name.

For organization purposes it might make sense to put all these argo-cd applications for the same repositry/directory into a single namespace.

Lets imagine the following directory structure:

- /cluster-1/cert-manager
- /cluster-1/external-dns
- /cluster-2/cert-manager
- /cluster-2/external-dns
- /cluster-3/cert-manager
- /cluster-3/external-dns

With the following ApplicationSet
```yaml
apiVersion: argoproj.io/v1alpha1
kind: ApplicationSet
metadata:
  name: applications
  namespace: argocd
spec:
  generators:
  - git:
      repoURL: ...
      revision: main
      directories:
      - path: **/*
  template:
    metadata:
      name: '{{path[0]}}'
      namespace: '{{path[1]}}' # this is currently not used, it always uses namespace of set
    spec:
      project: default
      source:
        repoURL: ...
        targetRevision: main
        path: '{{path}}'
      destination:
        name: '{{path[0]}}'
        namespace: '{{path[1]}}'
```

Which I would expect to generate applications like:
```yaml
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: cluster-1
  namespace: cert-manager
spec:
  project: default
  source:
    repoURL: ...
    targetRevision: main
    path: ...
  destination:
    name: cluster-1
    namespace: cert-manager
```

# Proposal

When `.spec.template.metadata.namespace` is set on the ApplicationSet, use that, otherwise fallback to the ApplicationSet's namespace.

**Comment by rishabh625** [comment]
Hi , assigning this to myself , would soon submit the change

**Comment by crenshaw-dev** [comment]
I think we should consider having the ApplicationSet controller enforce constrains on the `namespace` field.

For example, ApplicationSets in namespaces matching `dev-*` may only create apps in that same namespace. This would allow non-admin users to safely create ApplicationSets in a restricted namespace or namespaces.

But of course, there should also be a mode allowing an ApplicationSet to deploy to multiple destination namespaces. And ideally a way to enforce both options with the same controller.

Are folks considering more advanced use cases which we should consider?

**Comment by rishabh625** [comment]
> Hey. @rishabh625 any updates on this? Really looking forward for this to be added.

working on it today.

**Comment by marshell0** [comment]
> 

Any update? or which version will include this fix?

**Comment by crenshaw-dev** [comment]
This does what I wanted (appset in any namespace, deploying to same namespace): https://github.com/argoproj/argo-cd/pull/12378

I think it's distinct from this request, which is an appset deploying to other namespaces. I think both are valid.

**Comment by callppatel** [comment]
When this fix will be available?

**Comment by via-justa** [comment]
Is there any update on that?

**Comment by reedjosh** [comment]
Bump, would love this feature.

**Comment by mirza-src** [comment]
any updates on this? Really want this feature

**Comment by crenshaw-dev** [comment]
imo we need a proposal for an AppSetProject feature in order to make this request viable. 

The whole point of the apps-in-any-namespace feature is to enable self-service of applications. Appset-within-a-namespace allows safe self-service of appsets within user-controlled namespaces.

But appsets-spanning-namespaces again makes the AppSet effectively an admin resource. If I can create an AppSet that can install an App in any namespace, I can install an app in the namespace holding admin-level apps (usually the `argocd` namespace).

To make an appsets-spanning-namespaces feature usable, we need some way for admins to restrict what their users' AppSets can do. Examples of restrictions they may need:

1) Which namespaces the AppSets may be installed in
1) Which namespace the AppSet may install Apps in
1) Which projects the AppSet's generated Apps may use
1) Which SCM APIs the AppSet's PR and SCM generators may access

We would need to decide how the currently pseudo-project-scoped AppSet resource would be handled by the API server in a system which includes an AppSet Project.

This is a very, very big feature and would require a proposal to be presented and accepted.

Is anyone up for writing the proposal?

**Comment by mate4st** [comment]
Interesting, our use case differs that we do not care about self service, but having a namespace per cluster with its applications purely for organizational purposes. This basically prevents us to use ApplicationSets as we would than need to put all applications in one namespace. 

Additionally I wonder, aren't ApplicationSet's considered to be an "admin" resource for other [reasons](https://argo-cd.readthedocs.io/en/release-2.13/operator-manual/applicationset/Security/) anyway?

**Comment by illrill** [comment]
> The whole point of the apps-in-any-namespace feature is to enable gitops-based self-service of applications for non-admins.

This is not entirely true anymore. [Argo CD Agent](https://argocd-agent.readthedocs.io/latest/) is built upon the applications-in-any-namespace feature. In short, Applications destined to a certain remote cluster **must** be created in a namespace named after the remote cluster's Argo CD Agent. In this case, to use ApplicationSet with cluster generator, I need the ApplicationSet to create Applications in _other_ namespaces.

Given the high demand for this capability, I'd argue that the choice between namespace boundary enforcement vs. cross-namespace flexibility should be in the hands of the user, not strictly limited by the product itself. And @mate4st pointed out, the [security](https://argo-cd.readthedocs.io/en/release-2.13/operator-manual/applicationset/Security/) recommendations already state that ApplicationSet should be in the hands of cluster admins, so it shouldn't be _that_ big of a deal.

How about placing it behind an opt-in feature flag along with a warning sign?

See also https://cloud-native.slack.com/archives/C07L5SX6A9J/p1758289030402449

**Comment by alexrecuenco** [comment]
I think a good first step might be simply allowing AppSets with the default AppProject in the root namespace (`argocd`) to create apps in other namespaces, since it already ought to have all the permissions to do so anyway.

I think that would cover a lot of the use cases

**Comment by safaci2000** [comment]
> I think a good first step might be simply allowing AppSets with the default AppProject in the root namespace (`argocd`) to create apps in other namespaces, since it already ought to have all the permissions to do so anyway.

100% that would address my use case. I'm separating apps by namespace to avoid name collisions. If we could allow this feature to run from argocd that would cover all of our needs.

**Comment by crenshaw-dev** [comment]
The default project is initially installed has broad permissions, but those may be restricted after installation. And giving AppSets in the controller namespace superpowers seems a bit too magic to me.

**Comment by crenshaw-dev** [comment]
> aren't ApplicationSet's considered to be an "admin" resource for other [reasons](https://argo-cd.readthedocs.io/en/release-2.13/operator-manual/applicationset/Security/) anyway?

Yes, _but_, each of the reasons has mitigations:

> ApplicationSets can create Applications under arbitrary [Projects](https://argo-cd.readthedocs.io/en/release-2.13/user-guide/projects/).

This is true unless the AppSet is restricted to creating Apps in namespaces which are only allowed by certain projects, i.e. the way AppSets-in-any-namespace works today.

> ApplicationSets can also quickly create an arbitrary number of Applications and just as quickly delete them.

Non-admins often have access to create and delete arbitrary numbers of Apps, so I don't think this is a unique reason to treat AppSets as admin-only resources.

> Finally, ApplicationSets can reveal privileged information. For example, the [git generator](https://argo-cd.readthedocs.io/en/release-2.13/operator-manual/applicationset/Generators-Git/) can read Secrets in the Argo CD namespace and send them to arbitrary URLs (e.g. URL provided for the api field) as auth headers.

Sensitive settings like this often have admin-level restrictions. For example, admins can allowlist SCM provider URLs.

tl;dr - AppSets are relatively high-privilege resources, but I think the best direction for the project is to make these resource more accessible to non-admins.

**Comment by alexrecuenco** [comment]
Right, not as a special magical rule, but as a "if this project I am part of already has global wildcard admin permissions and can deploy to anywhere, then there shouldn't be issues".

But it is true that AppSets are hard to integrate in safe ways generally
